include/ginkgo/core/base/batch_lin_op.hpp

```diff
@@ -0,0 +1,439 @@
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_BATCH_LIN_OP_HPP_
+#define GKO_PUBLIC_CORE_BASE_BATCH_LIN_OP_HPP_
+
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+
+#include <ginkgo/core/base/abstract_factory.hpp>
+#include <ginkgo/core/base/batch_lin_op_helpers.hpp>
+#include <ginkgo/core/base/dim.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/matrix_assembly_data.hpp>
+#include <ginkgo/core/base/matrix_data.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/log/logger.hpp>
+
+
+namespace gko {
+namespace batch {
+
+
+/**
+ * @addtogroup BatchLinOp
+ *
+ * @section batch_linop_concept Batched Linear operator as a concept
+ *
+ * A batch linear operator (BatchLinOp) forms the base class for all batched
+ * linear algebra objects. In general, it follows the same structure as the
+ * LinOp class, but has some crucial differences which make it not strictly
+ * representable through or with the LinOp class.
+ *
+ * A batched operator is defined as a set of independent linear operators which
+ * have no communication/information exchange between them. Therefore, any
+ * collective operations between the batches is not possible and not
+ * implemented. This allows for each batch to be computed and operated on in an
+ * embarrasingly parallel fashion.
+ *
+ * Similar to the LinOp class, the BatchLinOp also implements
+ * BatchLinOp::apply() methods which call the internal apply_impl() methods
+ * which the concrete BatchLinOp's have to implement.
+ *
+ * A key difference between the LinOp and the BatchLinOp classes is the storing
+ * of dimensions. BatchLinOp allows for storing non-equal objects in the
+ * batches and hence stores a batch_dim object instead of a dim object. The
+ * batch_dim object is optimized to store less amount of data when storing
+ * uniform batches.
+ *
+ * All size validation functions again verify first that the number of batches
+ * are conformant and that the dimensions in the corresponding batches
+ * themselves are also valid/conformant. Here too, optimizations for uniform
+ * batches have been added.
+ *
+ * @ref BatchLinOp
+ */
+class BatchLinOp : public EnableAbstractPolymorphicObject<BatchLinOp> {
+public:
+    /**
+     * Applies a batch linear operator to a batch vector (or a sequence of batch
+     * of vectors).
+     *
+     * Performs the operation x = op(b), where op is this batch linear operator.
+     *
+     * @param b  the input batch vector(s) on which the batch operator is
+     *           applied
+     * @param x  the output batch vector(s) where the result is stored
+     *
+     * @return this
+     */
+    BatchLinOp* apply(ptr_param<const BatchLinOp> b, ptr_param<BatchLinOp> x)
```

Comment:
I guess this has not been updated with the batch::MultiVector in mind. The question is do we need to have both apply functions, one applying to vectors and one applying to general linops?

Reply:
I think this is actually a more difficult problem. The multi-vector is templated, so we can't just use it as a parameter for the virtual functions. So we would probably need to introduce a non-templated base class for the multi-vector. The only other alternative I can think of right now would be using variants. But that might lead to more issues than it's worth.

Reply:
What do you think about circumventing this whole thing and having only the concrete LinOp/BatchLinOp handling the apply with `MultiVector` parameters ? I think that significantly simplifies things without too much code in the concrete classes.

Reply:
I'm not sure I follow. So the `BatchLinOp` would not have an apply at all?

Reply:
`BatchLinOp` has apply only to `BatchLinOp`, but the `apply(MultiVector<> b ... )` is only in the concrete classes, `batch::Dense`, `batch::Csr` etc.

Reply:
For me that seems very odd, since now there is a linear operator that can't be applied to vectors, which is the whole definition of a linear operator. Also, you could not use `BatchLinOp` in solvers or similar (although that might not be as important as for the `LinOp`).

Reply:
Will the `apply(BatchLinOp*, BatchLinOp*)` be needed in the future? I think an argument can also be made to drop that entirely, because it only has very limited use, eg. SpGEAM but only for CSR. So maybe just remove it and the BatchLinOp only has it's size.

Reply:
It would be used for GEMM, which exists for all formats, and maybe also for user created LinOps ?

Reply:
I think we wanted to make a clear distinction between Dense used as a vector and as a matrix, see #796
In that setting, GEMM should be a separate function, not an `apply`.

Reply:
No, since you can only pass a vector to apply, it cannot be used to call GEMM.

Reply:
Perhaps to get back to my point, I don't think we will have any meaning full implementation of `appy(BatchLinOp*, BatchLinOp*)`, so we can just remove it here.

Reply:
So your suggestion would be to have a separate `gemm`, `geam` in all matrix formats instead ?

Reply:
TBH, for me there is no difference between GEMM and `apply(MultiVector*, MultiVector*)`, or I'm misunderstanding something.

Reply:
> So your suggestion would be to have a separate gemm, geam in all matrix formats instead ?

No, my suggestion would be to only add specific functions if we need them and they are feasible to implement. Most sparse storage formats are not well-suited for implementing gemm or geam.

> TBH, for me there is no difference between GEMM and apply(MultiVector*, MultiVector*), or I'm misunderstanding something.

If I say GEMM I mean dense * dense, not dense * multivector. The latter can of course use GEMM internally.

Reply:
The underlying data might be the same (this is only the case for dense though), but the semantics are different and you usually have dense objects that you call gemm on, and if you want to use the multivector apply, then you will have to create new multivector objects for those. 

For SpGEMM and SpGEAM, that is not the case, and is not representable with the multivector apply.